Add documentation for CI/CD set up (#18330)
nvaccess/nvda: ci/README.md

# Continuous Integration with GitHub Actions

## Background

GitHub Actions builds the following types of NVDA installers through a CI/CD pipeline:

* Pull Request builds: Generated from the pull requests.
Pull requests initiated from `nvaccess/nvda` rather than a fork have extra permissions than standard PRs.
* Snapshot builds: Generated from pushes to master/beta/rc or branch names prefixed with `try-`.
These are signed and deployed to the NV Access server.
* Tagged builds: Generated from pushes to tags prefixed with `release-`.
Official beta, rc and stable releases.
These are signed and deployed to the NV Access server.

## Builds

### Build process

The build process is non-linear.
Some of these steps run concurrently.

1. Prepare source code and cache:
    1. Checkout NVDA repository with submodules.
    1. Install dependencies (or use cache).
    1. Set version and scons variables.
    1. Build NVDA source.

> 1. Build and test:
>     1. Run static tests
>     1. Build launcher
>     1. Install NVDA
>     1. Run systems tests

Reviewer: If these jobs aren't serial, they shouldn't be ordered list items.
```suggestion
1. Build and test:
    * Run static (unit) tests
    * System tests
      1. Build launcher
      1. Install NVDA
      1. Run system tests
```

1. Deploy (not fully active currently):

>     1. On tagged/snapshot builds, upload symbols to Mozilla
>     1. On beta branch builds, upload translation to Crowdin.
>     1. On snapshot builds, deploy to the server.
>     1. On release builds, publish the release in GitHub and deploy to the server.

Reviewer: Presumably these are also run in parallel.
```suggestion
    * On tagged/snapshot builds, upload symbols to Mozilla
    * On beta branch builds, upload translation to Crowdin.
    * On snapshot builds, deploy to the server.
    * On release builds, publish the release on GitHub and deploy to the server.
```

To enable, set:

* `CROWDIN_PROJECT_ID` as a variable.
* `CROWDIN_AUTH_TOKEN` as a secret.

### Code signing

NV Access signs snapshot/tagged builds with SignPath.

To enable, set:

* `API_SIGNING_TOKEN` as a secret with your SignPath token.

### Uploading symbols

NVDA uploads its build symbols to Mozilla to help them with debugging on snapshot/tagged builds.

To enable, set:

* `MOZILLA_SYMS_TOKEN` as a secret.
* `feature_uploadSymbolsToMozilla` as a variable with any non-empty string.

Generating this requires direct co-ordination with Mozilla.

### GitHub Environments

GitHub Environments are used to protect deployments of snapshot/tagged builds.

Create two GitHub Environments, one called `production`, the other `snapshot`.

#### production

Used for tagged releases of NVDA.
It's recommended to enable deployment protection rules so that a human can confirm the deployment of a built tagged release staged for deployment.
This is so any desired testing can get manually confirmed, and communications for the release can be prepared.

Configuration:

* Name: `production`
* Enable required reviewers: this ensures a human must confirm deployment of a staged release

> * Under "Deployment branches and tags", create a tag rule: `release-*`

Reviewer: Shouldn't this be restricted to `release-*` tags on `beta`/`rc` branches? Or do we restrict pushing `release-*` tags to only `beta` and `rc` elsewhere?

#### snapshot

Configuration:

* Name: `snapshot`
* Do not enable required reviewers: this is not needed for snapshots.
* Under "Deployment branches and tags", create these branch rules:
  * `master`
  * `beta`
  * `rc`
  * `try-**`

### Deployment webhook

Create a GitHub webhook to subscribe to snapshot/tagged builds of NVDA, and use it to deploy to a server.

Under events, only subscribe to the Deployments event.

Ensure a secret is set and SSL is enabled.

### VirusTotal scanning

> NV Access scans tagged builds with VirusTotal.

Reviewer: To match established style (GH won't let me submit a single comment with this suggestion, so this and the next comment go together)
```suggestion
NV Access scans tagged builds with VirusTotal.
To enable tagged builds, set:
```

Author: This isn't quite accurate. Setting the secret doesn't enable tagged builds. Setting the secret is required for performing tagged builds.

> Set `VT_API_KEY` as a secret to perform tagged builds.

Reviewer: ```suggestion
* `VT_API_KEY` as a secret.
```

The build process is non-linear.
Some of these steps run concurrently.

> * Prepare source code and cache:
>   * Checkout NVDA repository with submodules.
>   * Install dependencies (or use cache).
>   * Set version and scons variables.
>   * Build NVDA source.
> * Build and test:
>   * Run static tests
>   * Build launcher
>   * Install NVDA
>   * Run systems tests
> * Deploy (not fully active currently):
>   * On tagged/snapshot builds, upload symbols to Mozilla
>   * On beta branch builds, upload translation to Crowdin.
>   * On snapshot builds, deploy to the server.
>   * On release builds, publish the release on GitHub and deploy to the server.
> * Clean up build cache.

Reviewer: I think you may have missed the point I was making - using `<ul>` and `<ol>` can help to indicate the dependency between steps.

Presumably "Prepare source code and cache" has to be run before "Build and test:", which has to be run before "Deploy", which has to be run before "Clean up build cache"; while "Run static tests" can be run in parallel with "Build launcher", "Install NVDA" and "Run system tests". 

```suggestion
1. Prepare source code and cache:
   1. Checkout NVDA repository with submodules.
   1. Install dependencies (or use cache).
   1. Set version and scons variables.
   1. Build NVDA source.
1. Build and test:
   * Run unit tests
   * Create launcher and run system tests:
     1. Build launcher
     1. Install NVDA
     1. Run system tests
1. Deploy (not fully active currently):
   * On tagged/snapshot builds, upload symbols to Mozilla
   * On beta branch builds, upload translation to Crowdin.
   * On snapshot builds, deploy to the server.
   * On release builds, publish the release on GitHub and deploy to the server.
1. Clean up build cache.
```

This is not strictly accurate (cleaning up the cache I think can happen more-or-less in parallel with the deploy jobs), but it gives a much clearer idea of how things run.

### Build number offset

To offset from our previous build system, we start the sequential build count at a higher number than 0.

> This means our first build will be numbered something like 100,001 not *

Reviewer: I do not understand this change?
```suggestion
This means our first build will be numbered something like 100,001 not 1.
```

Author: find and replace for `1.` stuffed up
